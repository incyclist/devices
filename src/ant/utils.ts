export const getBrand = (manId) => {
    if(manId===undefined)
        return "ANT+"

    switch (manId) {
        case 1: return "Garmin"; 
        case 16: return "Timex"; 
        case 23: return "Suunto";        
        case 52: return "Seiko";
        case 53: return "Seiko";
        case 70: return "Sigma";
        case 123: return "Polar"; 
        case 287: return "VDO"; 
        case 32: return "Wahoo"; 
        case 86: return "Elite"; 
        case 89: return "Tacx"; 
        default: 
            return "ANT+"
    }
}

