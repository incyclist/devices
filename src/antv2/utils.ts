import { Profile } from "incyclist-ant-plus"
import { LegacyProfile } from "./types"

const brands = [
    'Garmin,1',
    'garmin_fr405_antfs,2',
    'Zephyr,3',
    'Dayton,4',
    'idt,5',
    'SRM,6',
    'Quarq,7',
    'ibike,8',
    'Saris,9',
    'spark_hk,10',
    'Tanita,11',
    'echowell,12',
    'Dynastream_oem,13',
    'Nautilus,14',
    'Dynastream,15',
    'Timex,16',
    'Metrigear,17',
    'Xelic,18',
    'Beurer,19',
    'Cardiosport,20',
    'A&D,21',
    'Hmm,22',
    'Suunto,23',
    'Thita Elektronik,24',
    'Gpulse,25',
    'Clean Mobile,26',
    'Pedal Brain,27',
    'Peaksware,28',
    'Saxonar,29',
    'Lemond Fitness,30',
    'Dexcom,31',
    'Wahoo,32',
    'Octane,33',
    'Archinoetics,34',
    'The Hurt Box,35',
    'Citizen Systems,36',
    'Magellan,37',
    'Osynce,38',
    'Holux,39',
    'Concept2,40',
    'One Giant Leap,42',
    'Ace Sensor,43',
    'Brim Brothers,44',
    'Xplova,45',
    'Perception Digital,46',
    'Bf1systems,47',
    'Pioneer,48',
    'Spantec,49',
    'Metalogics,50',
    '4iiiis,51',
    'Seiko,52',
    'Seiko,53',
    'Ifor Powell,54',
    'Maxwell Guider,55',
    'Star rac,56',
    'Breakaway,57',
    'Alatech,58',
    'Mio,59',
    'Rotor,60',
    'Geonaute,61',
    'Id Bike,62',
    'Specialized,63',
    'Wtek,64',
    'Physical Enterprises,65',
    'North Pole,66',
    'Bkool,67',
    'Cateye,68',
    'Stages,69',
    'Sigma,70',
    'Tomtom,71',
    'Peripedal,72',
    'Wattbike,73',
    'Moxy,76',
    'Ciclosport,77',
    'Powerbahn,78',
    'Acorn,79',
    'Lifebeam,80',
    'Bontrager,81',
    'Wellgo,82',
    'Scosche,83',
    'Magura,84',
    'Woodway,85',
    'Elite,86',
    'Nielsen Kellerman,87',
    'DK City,88',
    'Tacx,89',
    'Direction,90',
    'Magtonic,91',
    '1partcarbon,92',
    'Inside Ride,93',
    'Sound Of Motion,94',
    'Stryd,95',
    'icg,96',
    'MiPulse,97',
    'Bsx Athletics,98',
    'Look,99',
    'Campagnolo,100',
    'Body Bike Smart,101',
    'Praxisworks,102',
    'Limits,103',
    'Topaction,104',
    'Cosinuss,105',
    'Fitcare,106',
    'Magene,107',
    'Giant,108',
    'Tigrasport,109',
    'Salutron,110',
    'Technogym,111',
    'Bryton,112',
    'Latitude,113',
    'Soaring,114',
    'Igpsport,115',
    'Thinkrider,116',
    'Gopher,117',
    'Waterrower,118',
    'Orangetheory,119',
    'Inpeak,120',
    'Kinetic,121',
    'Johnson,122',
    'Polar,123',
    'Seesense,124',
    'NCI,125',
    'Development,255',
    'Healthandlife,257',
    'Lezyne,258',
    'Scribe,259',
    'Zwift,260',
    'Watteam,261',
    'Recon,262',
    'Favero,263',
    'Dynovelo,264',
    'Strava,265',
    'Precor,266',
    'Bryton,267',
    'Sram,268',
    'Navman,269',
    'Cobi,270',
    'Spivi,271',
    'Mio Magellan,272',
    'Evesports,273',
    'Sensitivus Gauge,274',
    'Podoon,275',
    'Life Time Fitness,276',
    'Falco,277',
    'Minoura,278',
    'Cycliq,279',
    'Luxottica,280',
    'Trainer Road,281',
    'The Sufferfest,282',
    'Fullspeedahead,283',
    'Virtualtraining,284',
    'Feedbacksports,285',
    'Omata,286',
    'VDO,287',
    'Magneticdays,288',
    'Hammerhead,289',
    'Kinetic By Kurt,290',
    'Shapelog,291',
    'Dabuziduo,292',
    'Jetblack,293',
    'Coros,294',
    'Virtugo,295',
    'Velosense,296',
    'Cycligentinc,297',
    'Actigraphcorp,5759'
]

const brandLookup = brands.map( str => {
    const [name,id] = str.split(',')
    return {name,id:Number(id)}
})


export const getBrand = (manId:number) => {
    const found = brandLookup.find( l => l.id===manId)
    if (found)
        return found.name
    return;
}

export const mapLegacyProfile = ( legacy:LegacyProfile ):Profile => {
    switch (legacy){
        case 'Heartrate Monitor': return 'HR'
        case 'Power Meter': return 'PWR'
        case 'Smart Trainer': return 'FE'
        case 'Cadence Sensor':return 'CAD'
        case 'Speed Sensor': return 'SPD'
        case 'Speed + Cadence Sensor': return 'SC'
    }
} 

