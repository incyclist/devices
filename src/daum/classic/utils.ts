export function getCockpit( c) {
    switch ( c) {        
        case 10: 
            return "Cardio";
        case 20: 
            return "Fitness";
        case 30: 
            return "Vita De Luxe";
        case 40: 
            return "8008";
        case 50: 
            return "8080";
        case 60: 
            return "Therapie";
        default:
            return "Unknown";
    }
}

export function getBikeType( type) {
    if (type===undefined)
        return 0;

    switch (type) {
        case "triathlon":
            return 0;
        case "race":
            return 0;
        case "mountain":
            return 1;
        default:
            return 1;
    }
    
    // TODO: define possible values and parse (should be aligned with ANT* spec)
}


export function getAge(birthday) { // birthday is a date
    if (birthday===undefined) {
        return 30;
    }

    try {
        const bd = new Date(birthday);
        const ageDifMs = Date.now() - bd.getTime();
        var ageDate = new Date(ageDifMs); // miliseconds from epoch
        return Math.abs(ageDate.getUTCFullYear() - 1970);    
    } 
    catch (error) {
        return 30;
    }
}

export function getGender(sex) {
    if (sex===undefined)
        return 2;
    switch (sex) {
        case "M":
            return 1; 
        case "F":
            return 1; 
        default:
            return 2;
    }
    
}

export function getLength( length) {
    if (length===undefined || length===null)
        return 180; // average european

    let l = Math.round(length);

    if (l<100)
        return 100;
    if (l>220)
        return 220;

    return l;
}

export function getWeight(weight?) {
    if (weight===undefined || weight===null)
        return 80; // average european

    let m = Math.round(weight);
    if (isNaN(m))
        return 80;

    if (m<10)
        return 10;
    if (m>250)
        return 250;

    return m;
}

export function parseRunData( data) {
    const bikeData = {} as any
    bikeData.isPedalling = (data[4]>0);
    bikeData.power  = data[5]*5;
    bikeData.cadence = data[6];
    bikeData.speed = data[7];
    bikeData.heartrate = data[14];
    bikeData.distance = (data[9]*256+data[8])/10;
    bikeData.distanceInternal = (data[9]*256+data[8])*100;
    bikeData.time = (data[11]*256+data[10]);
    bikeData.gear = data[16];
    return bikeData
}


export function buildError ( status,err) {
    const message = err && err.message ? err.message: err;
    const error = new Error( message) as any;
    error.status = status;
    return error;
}

export function hexstr(arr,start?,len?) {
    var str = "";
    if (start===undefined) 
        start = 0;
    if ( len===undefined) {
        len = arr.length;
    }
    if (len-start>arr.length) {
        len = arr.length-start;
    }

    var j=start;
    for (var i = 0; i< len; i ++) {
        var hex = Math.abs( arr[j++]).toString(16);
        if ( i!==0 ) str+=" ";
        str+=hex;
    }
	return str;
}


export function Float32ToHex (float32)  {
    function getHex (i) { return ('00' + i.toString(16)).slice(-2) }
    var view = new DataView(new ArrayBuffer(4))
    view.setFloat32(0, float32);
    return Array.apply(null, { length: 4 }).map((_, i) => getHex(view.getUint8(i))).join('');
}

export function Float32ToIntArray (float32)  {
    var view = new DataView(new ArrayBuffer(4))
    view.setFloat32(0, float32);
    var arr = [];
  	for ( let i=0;i<4;i++) {
      arr.push( view.getUint8(i))
    }
    return arr;
}
