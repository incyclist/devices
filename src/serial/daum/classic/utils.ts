import {DeviceType,IncyclistBikeData} from '../../../types'

export const DEFAULT_AGE         = 30;
export const DEFAULT_USER_WEIGHT = 75;
export const DEFAULT_BIKE_WEIGHT = 10;

export function getCockpit(c:number) {
    switch ( c) {        
        case 10: 
            return "Cardio";
        case 20: 
            return "Fitness";
        case 30: 
            return "Vita De Luxe";
        case 40: 
            return "8008";
        case 0x2A:
            return "8008 TRS"
        case 50: 
            return "8080";
        case 60: 
            return "Therapie";
        case 100: 
            return "8008 TRS Pro";
        case 160: 
            return "8008 TRS3";
        case 0x8D: // 141
            return "ergo_lyps Cardio Pro"
        default:
            return "Unknown";
    }
}

export function getSerialNo(arr:Uint8Array, start:number, length:number) {
  
    const buffer = Buffer.from(arr.subarray(start,start+length))
    return buffer.toString('hex')
}

export function getBikeType(type?:DeviceType):number {

    const DAUM_CLASSIC_BT_MOUNTAIN = 1
    const DAUM_CLASSIC_BT_RACE = 0
    //const DAUM_CLASSIC_BT_ALLROUND = 2

    if (type===undefined)
        return DAUM_CLASSIC_BT_RACE;

    switch (type) {
        case 'triathlon':
            return DAUM_CLASSIC_BT_RACE;
        case 'race':
            return DAUM_CLASSIC_BT_RACE;
        case 'mountain':
            return DAUM_CLASSIC_BT_MOUNTAIN;
    }
   
}



export function getGender(sex) {
    if (sex===undefined)
        return 2;
    switch (sex) {
        case "M":
            return 0; 
        case "F":
            return 1; 
        default:
            return 2;
    }
    
}

export function getLength( length?:number) {
    if (length===undefined || length===null)
        return 180; // average european

    const l = Math.round(length);
    return between(l,100,220)
}

export function getWeight(weight?:number) {
    if (weight===undefined || weight===null)
        return 80; // average european

    let m = Math.round(weight);
    return between(m,10,250)
}

export function parseRunData( data) {
    const bikeData = {} as IncyclistBikeData


    /*
    const pedalling = data[4];
    if (pedalling===0x40 || pedalling===0x41) {
        throw new Error('Invalid data');
    }
    */

    bikeData.isPedalling = (data[4]>0);
    bikeData.power  = data[5]*5;
    bikeData.pedalRpm = data[6];
    bikeData.speed = data[7];
    bikeData.heartrate = data[14];
    bikeData.distanceInternal = (data[9]*256+data[8])*100;
    bikeData.time = (data[11]*256+data[10]);
    bikeData.gear = data[16];

    // TODO: calories

    return bikeData
}

export function between(v:number, min:number, max:number) {
    if (v<min)
        return min;
    if (v>max)
        return max;
    return v;
}

export function buildSetSlopeCommand(bikeNo,slope) {
    const buffer = Buffer.from([0x55,bikeNo,0,0,0,0]);
    buffer.writeFloatLE(slope,2);
    const cmd = Array.from(buffer)
    
    return cmd;
}

