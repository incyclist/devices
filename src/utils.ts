export const sleep = (ms) => { 
    return new Promise( resolve => setTimeout(resolve, ms))
};

export function runWithRetries( fn, maxRetries, timeBetween) {

    return new Promise( (resolve,reject)=> {
        let retries = 0;
        let tLastFailure = undefined;
        let busy = false;
        const iv = setInterval ( async ()=> {
            const tNow =Date.now();

            /* istanbul ignore next */
            if(busy) 
                return;

            if ( tLastFailure===undefined || tNow-tLastFailure>timeBetween) {
                try {
                    busy = true;
                    const data = await fn();
                    busy = false;
                    clearInterval(iv)
                    return resolve( data )
                }
                catch( err) {
                    tLastFailure = Date.now();
                    retries++;
                    if ( retries>=maxRetries) {
                        clearInterval(iv)
                        busy = false;
                        return reject( err);    
                    }         
                    else {
                        busy = false;
                    }
                }    
            }
        }, 50)
    })
}

export function hexstr(arr,start?,len?) {
    let str = "";

    const startIdx = start || 0;
    const length = len || arr.length;
    let endIdx = startIdx+length;
    if (endIdx>=arr.length) {
        endIdx = arr.length;
    }

    let added = 0;
    for (var i = startIdx; i< endIdx; i ++) {
        const hex = Math.abs( arr[i]).toString(16).toUpperCase();
        if ( added!==0 ) str+=' ';
        str+=hex;
        added++
    }
	return str;
}



export class Queue<T> {
    data: Array<T>;

    constructor( values? : Array<T>) {
        this.data = [];
        if (values)
            this.data = values;
    }

    size() {
        return this.data? this.data.length : 0;
    }

    clear() {
        this.data = [];
    }

    isEmpty() {
        return this.data===undefined || this.data.length===0;
    }

    dequeue(): T {
        const removed = this.data.splice(0,1)
        return removed[0];
    }

    enqueue(value: T) {
        this.data.push(value)
    }

}

