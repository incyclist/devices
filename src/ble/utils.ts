import { EventLogger } from "gd-eventlog";
import { LegacyProfile } from "../antv2/types";
import { BleCharacteristic,  BleProperty,  BleProtocol, BleRawPeripheral } from "./types";

type MappingRecord = {
    profile: LegacyProfile,
    protocol: BleProtocol,
}

export function mapLegacyProfile(profile:string):MappingRecord {
    switch (profile) {
        case 'Smart Trainer': return { profile:'Smart Trainer', protocol:'fm' }
        case 'Elite Smart Trainer': return { profile:'Smart Trainer', protocol:'elite' }
        case 'Elite SmartTrainer': return { profile:'Smart Trainer', protocol:'elite' }
        case 'Heartrate Monitor': return { profile:'Heartrate Monitor', protocol:'hr' }
        case 'Power Meter': return { profile:'Power Meter', protocol:'cp' }
        case 'Tacx Smart Trainer': return { profile:'Smart Trainer', protocol:'tacx' }
        case 'Tacx SmartTrainer': return { profile:'Smart Trainer', protocol:'tacx' }
        case 'Wahoo SmartTrainer': return { profile:'Smart Trainer', protocol:'wahoo' }
        case 'Wahoo Smart Trainer': return { profile:'Smart Trainer', protocol:'wahoo' }
    }
}

export function uuid (s:string):string {
    //console.log(s)
    if (s) {
        if (s.includes('-')) {
            const parts = s.split('-')
            const uuidNo = parseInt('0x'+parts[0])
            return uuidNo.toString(16).toLowerCase()
        }
        return s;
    }
}

export function matches (uuid1:string,uuid2:string):boolean {

    return parseUUID(uuid1)===parseUUID(uuid2)

}

export function getPeripheralInfo(p: BleRawPeripheral) {
    const {id,name,address,advertisement,services} = p;
    if (advertisement) {
        return {id,name:advertisement.localName,address,services:advertisement.serviceUuids}
    }
    else {
        return {id,name,address,services}
    }
}

export function getCharachteristicsInfo(c:BleCharacteristic) {
    const {uuid,properties,name,_serviceUuid} = c;

    const nameStr= name ? ` (${name})` : ''
    const serviceStr = _serviceUuid ? `${_serviceUuid}:` : ''

    return `${serviceStr}${uuid}${nameStr} ${properties}`
}




export const parseUUID = (str:string):string => {
    const uuid = str.toUpperCase()

    if (str.startsWith('0x')) { 
        const hex = uuid.slice(2)
        if (hex.length===4) { // 16bit)
            return `0000${hex}00001000800000805F9B34FB`    
        }
        if (hex.length===8) { // 32bit)
            return `${hex}00001000800000805F9B34FB`    
        }
    }
    else if (uuid.length===4) {
        return `0000${uuid}00001000800000805F9B34FB`
    }
    else if (uuid.length===8) {
        return `${uuid}00001000800000805F9B34FB`
    }
    else if (uuid.length===32) {
        return uuid
    }
    else if (uuid.length===36) {
        return uuid.replace(/-/g,'')

    }
    throw new Error(`Invalid UUID: ${uuid}`)
}

export const beautifyUUID = (str:string, withX:boolean = false ):string => {   

    let uuid

    try {
        uuid = parseUUID(str)
    }
    catch (err) {
        const logger = new EventLogger('Incyclist')
        logger.logEvent({message:'beautifyUUID error',uuid:str, error:err.message})
        return str
    }
    
    const parts = [
        uuid.substring(0,8),
        uuid.substring(8,12),    
        uuid.substring(12,16),
        uuid.substring(16,20),
        uuid.substring(20),
    ]

    if (uuid.substring(8)==='00001000800000805F9B34FB') {
        let short
        if (parts[0].startsWith('0000')) {
            short = parts[0].substring(4)
        }
        else {
            short  = parts[0]
        }
        return withX ? `0x${short}` : short
    }
    
    return parts.join('-')

}

export const fullUUID = (str:string):string => {   

    const uuid = parseUUID(str)
    
    const parts = [
        uuid.substring(0,8),
        uuid.substring(8,12),    
        uuid.substring(12,16),
        uuid.substring(16,20),
        uuid.substring(20),
    ]  
    return parts.join('-')

}


export const propertyVal = (properties:BleProperty[]):number => {
    let res = 0

    if (properties.includes('read'))    res |= 0x01
    if (properties.includes('write'))   res |= 0x02
    if (properties.includes('notify'))  res |= 0x04

    return res
}


export const propertyFromVal = (val:number):BleProperty[] => {
    const res:BleProperty[] = []

    if (val & 0x01) res.push('read')
    if (val & 0x02) res.push('write')   
    if (val & 0x04) res.push('notify')

    return res
}