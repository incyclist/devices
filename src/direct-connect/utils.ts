import { BleProperty } from "../ble/types"

export const parseUUID = (str:string):string => {
    const uuid = str.toUpperCase()

    if (str.startsWith('0x')) { 
        const hex = uuid.slice(2)
        if (hex.length===4) { // 16bit)
            return `0000${hex}00001000800000805F9B34FB`    
        }
        if (hex.length===8) { // 32bit)
            return `${hex}00001000800000805F9B34FB`    
        }
    }
    else if (uuid.length===4) {
        return `0000${uuid}00001000800000805F9B34FB`
    }
    else if (uuid.length===8) {
        return `${uuid}00001000800000805F9B34FB`
    }
    else if (uuid.length===32) {
        return uuid
    }
    else if (uuid.length===36) {
        return uuid.replace(/-/g,'')

    }
    throw new Error(`Invalid UUID: ${uuid}`)
}

export const beautifyUUID = (str:string, withX:boolean = false ):string => {   

    const uuid = parseUUID(str)
    
    const parts = [
        uuid.substring(0,8),
        uuid.substring(8,12),    
        uuid.substring(12,16),
        uuid.substring(16,20),
        uuid.substring(20),
    ]

    if (uuid.substring(8)==='00001000800000805F9B34FB') {
        let short
        if (parts[0].startsWith('0000')) {
            short = parts[0].substring(4)
        }
        else {
            short  = parts[0]
        }
        return withX ? `0x${short}` : short
    }
    
    return parts.join('-')

}

export const propertyVal = (properties:BleProperty[]):number => {
    let res = 0

    if (properties.includes('read'))    res |= 0x01
    if (properties.includes('write'))   res |= 0x02
    if (properties.includes('notify'))  res |= 0x04

    return res
}


export const propertyFromVal = (val:number):BleProperty[] => {
    const res:BleProperty[] = []

    if (val & 0x01) res.push('read')
    if (val & 0x02) res.push('write')   
    if (val & 0x04) res.push('notify')

    return res
}
