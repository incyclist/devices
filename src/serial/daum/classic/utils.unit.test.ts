
import { buildSetSlopeCommand, getBikeType, getCockpit, getGender, getWeight, parseRunData } from "./utils"

describe ('utils',()=>{

    describe('getCockpit',()=>{
        const res:string [] = []
        for (let i=0;i<256;i++)
            res.push( `0x${i.toString(16)}:${getCockpit(i)}` )
        expect(res).toMatchSnapshot()
    })

    describe('getBikeType',()=>{
        const res:string [] = []
        
        
        res.push( `race:${getBikeType('race')}` )
        res.push( `mountain:${getBikeType('mountain')}` )
        res.push( `triathlon:${getBikeType('triathlon')}` )
        res.push( `undefined:${getBikeType()}` )

        expect(res).toMatchSnapshot()
    })

    describe('getGender',()=>{        
        const testSet = ['M','F','X',undefined]
        const res = testSet.map( (i) =>  `0x${i}:${getGender(i)}` )
        expect(res).toMatchSnapshot()
    })

    describe('getWeight',()=>{
        test( '50kg',()=>{
            const res=getWeight(50)
            expect(res).toBe(50)
        })
    
        test( '50.123kg => returns 50',()=>{
            const res=getWeight(50.123)
            expect(res).toBe(50)
        })
    
        test( '49.5kg => returns 50',()=>{
            const res=getWeight(49.5)
            expect(res).toBe(50)
        })
    
        test( '<10kg => returns 10',()=>{
            const res=getWeight(2)
            expect(res).toBe(10)
        })
    
        test( '<250kg => returns 250',()=>{
            const res=getWeight(2000)
            expect(res).toBe(250)
        })
    
        test( 'undefined => returns 80',()=>{
            const res=getWeight(undefined)
            expect(res).toBe(80)
        })
        test( 'null => returns 80',()=>{
            const res=getWeight(null)
            expect(res).toBe(80)
        })
   
    
    
    })

    describe ('parseRunData',()=> { 
        test( 'valid data' ,()=>{ 
            let error=undefined;
            let data;
            try {
                data = parseRunData( [0x40,0,0,0,0xC0,0x13,0,0,0xF9,3,0x28,0x2E,3,41,0,0,7,0xE4,0x32]);
            } catch (e) {
                error =e;
            }
            expect(error).toBeUndefined();
            expect(data).toBeDefined()
            expect(data).toMatchObject( {power:95})
        })

        test( 'invalidvalid data: "pedalling=0x40"' ,()=>{ 
            let error=undefined;
            let data;
            try {
                data = parseRunData( [0x40,0,0,0,0x40,5,0,0,0,0,0,0,0,0,0,0,0xA,0,0]);
            } catch (e) {
                error =e;
            }
            expect(error).toBeUndefined();
            expect(data).toBeDefined()
            expect(data).toMatchObject( {power:25})

            //expect(error).toBeDefined();
            //expect(error.message).toBe('Invalid data')
        })
    })

    describe('buildSetSlopeCommand',()=>{
        test('0',()=>{
            const cmd = buildSetSlopeCommand(0,0)
            const res = Buffer.from(cmd).toString('hex').toUpperCase()
            
            expect(res).toBe('550000000000')
        } )

        test('88.5',()=>{
            const cmd = buildSetSlopeCommand(0,88.5)
            const res = Buffer.from(cmd).toString('hex').toUpperCase()
            expect(res).toBe('55000000B142')
        } )

        test('2.4',()=>{
            const cmd = buildSetSlopeCommand(0,2.4)
            const res = Buffer.from(cmd).toString('hex').toUpperCase()
            expect(res).toBe('55009A991940')
        } )

        test('3.3',()=>{
            const cmd = buildSetSlopeCommand(0,3.3)
            const res = Buffer.from(cmd).toString('hex').toUpperCase()
            expect(res).toBe('550033335340')
        } )

    })

})