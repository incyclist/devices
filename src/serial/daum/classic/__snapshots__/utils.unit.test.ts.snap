// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[` 1`] = `
[
  "0x0:Unknown",
  "0x1:Unknown",
  "0x2:Unknown",
  "0x3:Unknown",
  "0x4:Unknown",
  "0x5:Unknown",
  "0x6:Unknown",
  "0x7:Unknown",
  "0x8:Unknown",
  "0x9:Unknown",
  "0xa:Cardio",
  "0xb:Unknown",
  "0xc:Unknown",
  "0xd:Unknown",
  "0xe:Unknown",
  "0xf:Unknown",
  "0x10:Unknown",
  "0x11:Unknown",
  "0x12:Unknown",
  "0x13:Unknown",
  "0x14:Fitness",
  "0x15:Unknown",
  "0x16:Unknown",
  "0x17:Unknown",
  "0x18:Unknown",
  "0x19:Unknown",
  "0x1a:Unknown",
  "0x1b:Unknown",
  "0x1c:Unknown",
  "0x1d:Unknown",
  "0x1e:Vita De Luxe",
  "0x1f:Unknown",
  "0x20:Unknown",
  "0x21:Unknown",
  "0x22:Unknown",
  "0x23:Unknown",
  "0x24:Unknown",
  "0x25:Unknown",
  "0x26:Unknown",
  "0x27:Unknown",
  "0x28:8008",
  "0x29:Unknown",
  "0x2a:8008 TRS",
  "0x2b:Unknown",
  "0x2c:Unknown",
  "0x2d:Unknown",
  "0x2e:Unknown",
  "0x2f:Unknown",
  "0x30:Unknown",
  "0x31:Unknown",
  "0x32:8080",
  "0x33:Unknown",
  "0x34:Unknown",
  "0x35:Unknown",
  "0x36:Unknown",
  "0x37:Unknown",
  "0x38:Unknown",
  "0x39:Unknown",
  "0x3a:Unknown",
  "0x3b:Unknown",
  "0x3c:Therapie",
  "0x3d:Unknown",
  "0x3e:Unknown",
  "0x3f:Unknown",
  "0x40:Unknown",
  "0x41:Unknown",
  "0x42:Unknown",
  "0x43:Unknown",
  "0x44:Unknown",
  "0x45:Unknown",
  "0x46:Unknown",
  "0x47:Unknown",
  "0x48:Unknown",
  "0x49:Unknown",
  "0x4a:Unknown",
  "0x4b:Unknown",
  "0x4c:Unknown",
  "0x4d:Unknown",
  "0x4e:Unknown",
  "0x4f:Unknown",
  "0x50:Unknown",
  "0x51:Unknown",
  "0x52:Unknown",
  "0x53:Unknown",
  "0x54:Unknown",
  "0x55:Unknown",
  "0x56:Unknown",
  "0x57:Unknown",
  "0x58:Unknown",
  "0x59:Unknown",
  "0x5a:Unknown",
  "0x5b:Unknown",
  "0x5c:Unknown",
  "0x5d:Unknown",
  "0x5e:Unknown",
  "0x5f:Unknown",
  "0x60:Unknown",
  "0x61:Unknown",
  "0x62:Unknown",
  "0x63:Unknown",
  "0x64:8008 TRS Pro",
  "0x65:Unknown",
  "0x66:Unknown",
  "0x67:Unknown",
  "0x68:Unknown",
  "0x69:Unknown",
  "0x6a:Unknown",
  "0x6b:Unknown",
  "0x6c:Unknown",
  "0x6d:Unknown",
  "0x6e:Unknown",
  "0x6f:Unknown",
  "0x70:Unknown",
  "0x71:Unknown",
  "0x72:Unknown",
  "0x73:Unknown",
  "0x74:Unknown",
  "0x75:Unknown",
  "0x76:Unknown",
  "0x77:Unknown",
  "0x78:Unknown",
  "0x79:Unknown",
  "0x7a:Unknown",
  "0x7b:Unknown",
  "0x7c:Unknown",
  "0x7d:Unknown",
  "0x7e:Unknown",
  "0x7f:Unknown",
  "0x80:Unknown",
  "0x81:Unknown",
  "0x82:Unknown",
  "0x83:Unknown",
  "0x84:Unknown",
  "0x85:Unknown",
  "0x86:Unknown",
  "0x87:Unknown",
  "0x88:Unknown",
  "0x89:Unknown",
  "0x8a:Unknown",
  "0x8b:Unknown",
  "0x8c:Unknown",
  "0x8d:ergo_lyps Cardio Pro",
  "0x8e:Unknown",
  "0x8f:Unknown",
  "0x90:Unknown",
  "0x91:Unknown",
  "0x92:Unknown",
  "0x93:Unknown",
  "0x94:Unknown",
  "0x95:Unknown",
  "0x96:Unknown",
  "0x97:Unknown",
  "0x98:Unknown",
  "0x99:Unknown",
  "0x9a:Unknown",
  "0x9b:Unknown",
  "0x9c:Unknown",
  "0x9d:Unknown",
  "0x9e:Unknown",
  "0x9f:Unknown",
  "0xa0:8008 TRS3",
  "0xa1:Unknown",
  "0xa2:Unknown",
  "0xa3:Unknown",
  "0xa4:Unknown",
  "0xa5:Unknown",
  "0xa6:Unknown",
  "0xa7:Unknown",
  "0xa8:Unknown",
  "0xa9:Unknown",
  "0xaa:Unknown",
  "0xab:Unknown",
  "0xac:Unknown",
  "0xad:Unknown",
  "0xae:Unknown",
  "0xaf:Unknown",
  "0xb0:Unknown",
  "0xb1:Unknown",
  "0xb2:Unknown",
  "0xb3:Unknown",
  "0xb4:Unknown",
  "0xb5:Unknown",
  "0xb6:Unknown",
  "0xb7:Unknown",
  "0xb8:Unknown",
  "0xb9:Unknown",
  "0xba:Unknown",
  "0xbb:Unknown",
  "0xbc:Unknown",
  "0xbd:Unknown",
  "0xbe:Unknown",
  "0xbf:Unknown",
  "0xc0:Unknown",
  "0xc1:Unknown",
  "0xc2:Unknown",
  "0xc3:Unknown",
  "0xc4:Unknown",
  "0xc5:Unknown",
  "0xc6:Unknown",
  "0xc7:Unknown",
  "0xc8:Unknown",
  "0xc9:Unknown",
  "0xca:Unknown",
  "0xcb:Unknown",
  "0xcc:Unknown",
  "0xcd:Unknown",
  "0xce:Unknown",
  "0xcf:Unknown",
  "0xd0:Unknown",
  "0xd1:Unknown",
  "0xd2:Unknown",
  "0xd3:Unknown",
  "0xd4:Unknown",
  "0xd5:Unknown",
  "0xd6:Unknown",
  "0xd7:Unknown",
  "0xd8:Unknown",
  "0xd9:Unknown",
  "0xda:Unknown",
  "0xdb:Unknown",
  "0xdc:Unknown",
  "0xdd:Unknown",
  "0xde:Unknown",
  "0xdf:Unknown",
  "0xe0:Unknown",
  "0xe1:Unknown",
  "0xe2:Unknown",
  "0xe3:Unknown",
  "0xe4:Unknown",
  "0xe5:Unknown",
  "0xe6:Unknown",
  "0xe7:Unknown",
  "0xe8:Unknown",
  "0xe9:Unknown",
  "0xea:Unknown",
  "0xeb:Unknown",
  "0xec:Unknown",
  "0xed:Unknown",
  "0xee:Unknown",
  "0xef:Unknown",
  "0xf0:Unknown",
  "0xf1:Unknown",
  "0xf2:Unknown",
  "0xf3:Unknown",
  "0xf4:Unknown",
  "0xf5:Unknown",
  "0xf6:Unknown",
  "0xf7:Unknown",
  "0xf8:Unknown",
  "0xf9:Unknown",
  "0xfa:Unknown",
  "0xfb:Unknown",
  "0xfc:Unknown",
  "0xfd:Unknown",
  "0xfe:Unknown",
  "0xff:Unknown",
]
`;

exports[` 2`] = `
[
  "race:0",
  "mountain:1",
  "triathlon:0",
  "undefined:0",
]
`;

exports[` 3`] = `
[
  "0xM:0",
  "0xF:1",
  "0xX:2",
  "0xundefined:2",
]
`;
